piet/src/text.rs

```diff
@@ -94,6 +135,26 @@ pub trait TextLayout {
     fn hit_test_text_position(&self, text_position: usize) -> Option<HitTestTextPosition>;
 }
 
+/// Metadata about each line in a text layout.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct LineMetric {
+    /// Index (in code units) of the start of the line, offset from the beginning of the text.
+    pub line_start_offset: usize,
+
+    /// Line length (in code units), but offset from the beginning of the text. So it's the length
+    /// of this line summed with the lengths of all the lines before it.
+    ///
+    /// Does not include trailing whitespace.
+    pub line_end_offset: usize,
+
+    /// Line width.
+    /// Does not include trailing whitespace.
+    pub cumulative_width: f64,
+
```

Comment:
we're also going to need a way to get the *height* of the line, as well as the position of the baseline relative to the height.

Ideally we would get both the nominative line height (that is, the value by which lines should be spaced) as well as the *actual* height of the line as rendered, by which I mean the height of the rect that encloses all of the glyphs in the line; these can be different in certain cases.

Reply:
This LineMetrics struct is designed to be an platform-neutral abstraction of [DWRITE_LINE_METRICS ](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/ns-dwrite-dwrite_line_metrics), which has height and baseline fields. These are enough to do basic formatting and alignment. I suggest we add them.

In addition to that, I'm proposing that we store the cumulative sum of these. That makes binary searching efficient, and it's also efficient to get metrics for one line by subtracting two adjacent lines (we can have a query function that does this as a convenience).

Reply:
I'm not sure how well that translates to CoreText, but I haven't looked too deeply.

I'm also not sure about the cumulative sum stuff, which feels like an implementation detail, not necessarily part of the api? But I'm not too concerned about the details, I just want the basics.